joblib/parallel.pyi

```diff
@@ -0,0 +1,122 @@
+from typing import Any, Callable, Dict, Generator, List, Optional, Tuple, TypeVar
+
+from ._multiprocessing_helpers import mp as mp
+from ._parallel_backends import AutoBatchingMixin as AutoBatchingMixin
+from ._parallel_backends import FallbackToBackend as FallbackToBackend
+from ._parallel_backends import LokyBackend as LokyBackend
+from ._parallel_backends import MultiprocessingBackend as MultiprocessingBackend
+from ._parallel_backends import ParallelBackendBase as ParallelBackendBase
+from ._parallel_backends import SequentialBackend as SequentialBackend
+from ._parallel_backends import ThreadingBackend as ThreadingBackend
+from .disk import memstr_to_bytes as memstr_to_bytes
+from .externals import loky as loky
+from .externals.cloudpickle import dumps as dumps
+from .externals.cloudpickle import loads as loads
+from .logger import Logger as Logger
+from .logger import short_format_time as short_format_time
+
+delayedFunctionReturnType = TypeVar("delayedFunctionReturnType")
+
+BACKENDS: Any
+DEFAULT_BACKEND: str
+DEFAULT_N_JOBS: int
+DEFAULT_THREAD_BACKEND: str
+VALID_BACKEND_HINTS: Any
+VALID_BACKEND_CONSTRAINTS: Any
+EXTERNAL_BACKENDS: Any
+
+def get_active_backend(
+    prefer: Optional[Any] = ..., require: Optional[Any] = ..., verbose: int = ...
+): ...
+
+class parallel_backend:
+    old_backend_and_jobs: Any = ...
+    new_backend_and_jobs: Any = ...
+    def __init__(
+        self,
+        backend: Any,
+        n_jobs: int = ...,
+        inner_max_num_threads: Optional[Any] = ...,
+        **backend_params: Any
+    ) -> None: ...
+    def __enter__(self): ...
+    def __exit__(self, type: Any, value: Any, traceback: Any) -> None: ...
+    def unregister(self) -> None: ...
+
+DEFAULT_MP_CONTEXT: Any
+method: Any
+
+class BatchedCalls:
+    items: Any = ...
+    def __init__(
+        self,
+        iterator_slice: Any,
+        backend_and_jobs: Any,
+        reducer_callback: Optional[Any] = ...,
+        pickle_cache: Optional[Any] = ...,
+    ) -> None: ...
+    def __call__(self): ...
+    def __reduce__(self): ...
+    def __len__(self): ...
+
+def cpu_count(only_physical_cores: bool = ...): ...
+
+# def delayed(function: Any): ...
+def delayed(
+    whattododelayed: Callable[..., delayedFunctionReturnType]
+) -> Callable[
+    ..., Tuple[Callable[..., delayedFunctionReturnType], Tuple, Dict[str, Any]]
+]: ...
+
+class BatchCompletionCallBack:
+    dispatch_timestamp: Any = ...
+    batch_size: Any = ...
+    parallel: Any = ...
+    def __init__(
+        self, dispatch_timestamp: Any, batch_size: Any, parallel: Any
+    ) -> None: ...
+    def __call__(self, out: Any) -> None: ...
```

Comment:
I answered in the long text: For my understanding type hinting should be done at the "lowest layer" - the upper ones could be derived from the lowers.
Further I think it might be wrong to place type hints (for example) for `delayed` into __init__.pyi and it is not defined in __init__.py but only referenced.